feat: notify on reactions

VenkataRohan opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->

<!-- Your Pull Request name should start with one of the following tags
  feat: Adding a new feature
  refactor: A code change that doesn't change behavior (it doesn't add anything and doesn't fix anything)
  fix: For bug fixes that affect the end-user
  chore: For small tasks
  docs: For documentation
  ci: For updating CI configuration
  test: For adding tests
  i18n: For updating any translations
  regression: Issues created/reported/fixed during the development phase. kind of problem that never existed in production and that we don't need to list in a changelog for the end user
-->

<!-- Checklist!!! If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. 
  - I have read the Contributing Guide - https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat doc
  - I have signed the CLA - https://cla-assistant.io/RocketChat/Rocket.Chat
  - Lint and unit tests pass locally with my changes
  - I have added tests that prove my fix is effective or that my feature works (if applicable)
  - I have added necessary documentation (if applicable)
  - Any dependent changes have been merged and published in downstream modules
-->

## Proposed changes (including videos or screenshots)

isssue  #32297 

This PR notifies the user when someone reacts on the message and also increases the unread count in the sidebar

<img width="1535" alt="rocket_chat" src="https://github.com/RocketChat/Rocket.Chat/assets/83659881/0a6bf6a3-e213-468e-b32c-d5680e0a5b66">

<!--
  Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
  If it fixes a bug or resolves a feature request, be sure to link to that issue below.
  This description won't be displayed to our end users in the release notes, so feel free to add as much technical context as needed.
  If the changes introduced in this pull request must be presented in the release notes, make sure to add a changeset file. Check our guidelines for adding a changeset to your pull request: https://developer.rocket.chat/contribute-to-rocket.chat/modes-of-contribution/participate-in-rocket.chat-development/development-workflow#4.-adding-changeset-to-your-pull-request 
-->

## Issue(s)
<!-- Link the issues being closed by or related to this PR. For example, you can use #594 if this PR closes issue number 594 -->

## Steps to test or reproduce
<!-- Mention how you would reproduce the bug if not mentioned on the issue page already. Also mention which screens are going to have the changes if applicable -->

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

VenkataRohan commented:
Hi @sampaiodiego, I'm new to open source contributions and have recently developed a feature based on an issue. I submitted a PR a few days ago. Could you please suggest what my next steps should be to ensure my PR gets merged successfully?

VenkataRohan commented:
Hi @MarcosSpessatto  , I'm new to open source contributions and have recently developed a feature based on an issue. I submitted a PR a few days ago. Could you please suggest what my next steps should be to ensure my PR gets merged successfully?

VenkataRohan commented:
Hi @reetp  , I'm new to open source contributions and have recently developed a feature based on an issue. I submitted a PR a few days ago. Could you please suggest what my next steps should be to ensure my PR gets merged successfully?

reetp commented:
You don't need to ping me.

You have the attention of a dev.

Keep your PR in line with the latest code and hope they choose to merge it.

That will likely be up to the product team.

An "Enable/Disable" preference might be worth looking at too - not everyone wants to be bombarded with notifications!

casalsgh commented:
@VenkataRohan can you make it availabe as a possible setting?
We don't want to change the default behavior we have, but also want to permit this to be a possible setup.

VenkataRohan commented:
Hi @casalsgh , sure I am working on it , I will update here as soon as I implement it

ggazzo reviewed (changes requested):
I think its totally fine doing this with no setting.

what I disagree is to increment the unread, let keep this as ephemeral

VenkataRohan committed:
feat:notify on reation as optional setup

VenkataRohan commented:
Hi @ggazzo @casalsgh , now in this commit I have removed increase unread count and added an option of selecting notification on reaction
![image](https://github.com/RocketChat/Rocket.Chat/assets/83659881/d996fc58-96e1-4ee2-9dd3-52543a610d34)

ggazzo commented:
hey @VenkataRohan thanks for all the changes. lets keep the setting so, its not a big deal.

ggazzo commented on apps/meteor/lib/utils/parseReaction.ts:
```diff
@@ -0,0 +1,14 @@
+export function parseReaction(reactionText: string, roomtype: RoomType): string {
+	let reaction = '';
+	if (roomtype === 'd') {
+		reaction = `reacted with ${getEmojiClassNameAndDataTitle(reactionText.trim()).children}`;
```
what if we just use the reaction name and build the message on the clients?

anyway the translation is missing

VenkataRohan replied:
Hi @ggazzo ,
I have been working on a translation task to obtain reaction names from the back-end and build messages on the client-side. To achieve this, changes need to be made to one of the following files:

[konchatNotifications.ts](https://github.com/RocketChat/Rocket.Chat/blob/develop/apps/meteor/client/startup/notifications/konchatNotifications.ts) at line 86, where we receive the notification payload from the back-end.
[KonchatNotification.ts](https://github.com/RocketChat/Rocket.Chat/blob/develop/apps/meteor/app/ui/client/lib/KonchatNotification.ts) at line 36, where the above file calls this function.

However, since useTranslation is a hook and cannot be used in either of these files as they lack React components, I devised an alternative approach. Though it is somewhat verbose, it successfully translates the messages. I will make a commit with this working approach , please take a look.


<img width="334" alt="Screenshot 2024-06-22 at 10 30 22 PM" src="https://github.com/RocketChat/Rocket.Chat/assets/83659881/dd1e51e2-6bb1-4ad9-a735-32a0e517b885">

VenkataRohan commented on apps/meteor/app/ui/client/lib/KonchatNotification.ts:
```diff
@@ -56,7 +58,7 @@ class KonchatNotification {
 		const n = new Notification(notification.title, {
 			icon: notification.icon || getUserAvatarURL(notification.payload.sender?.username as string),
-			body: stripTags(message.msg),
+			body: notification.reacted ? parseReaction(notification.text, notification.payload.type as RoomType) : stripTags(message.msg),
```
we are actually sending reaction text from the backend and parsing it in the client.

VenkataRohan committed:
fix: Translation of reacted with

Akshanshkaushal commented:
I would like to work on this issue if it is not closed.

VenkataRohan commented:
> I would like to work on this issue if it is not closed.
Hi @Akshanshkaushal  I was waiting for a reply from @casalsgh , @ggazzo , I have already implement the feature and made changes suggested , waiting for the team to review it , and ready to work on any suggestion if required.

reetp commented:
> Hi @ggazzo @casalsgh , now in this commit I have removed increase unread count and added an option of selecting notification on reaction

Good job. These sorts of things are like Marmite. Some love them, some hate them.

I'd guess it could chew through a lot of push notifications too?

This could drive many people mad, especially in busy channels.

So options to disable it are good.

Parvezkhan0 commented on apps/meteor/app/lib/server/functions/notifications/desktop.ts:
```diff
@@ -22,21 +22,27 @@ export async function notifyDesktopUser({
 	const payload = {
 		title: title || '',
 		text,
+		reactionWithTranslation,
+		reacted: reaction !== '',
```
The reactionWithTranslation field is introduced but does not seem to have a default value or fallback mechanism when the translation fails. Ensure that reactionWithTranslation gracefully defaults to the untranslated reaction string to prevent any undefined behavior or empty notifications.

VenkataRohan replied:
Hi @Parvezkhan0 ,Sure I am working on it

VenkataRohan committed:
Merge branch 'develop' of https://github.com/RocketChat/Rocket.Chat into feat-notify-reactions

VenkataRohan committed:
Merge branch 'develop' of https://github.com/RocketChat/Rocket.Chat into feat-notify-reactions

VenkataRohan committed:
feat: added message to which user reacted along with emoji in notification

VenkataRohan commented:
Hi @Parvezkhan0, @ggazzo , @casalsgh 

I have added added fallback mechanism if translation fails as request by @Parvezkhan0 and also some good amount for changes to the code to optimse the translation of "Reacted With" process 

also I have added the message to which in the notification as shown below. 

the latest commit a9a40528588099d2b4bed9c39c7cdae09da3e89d has all the change. we can also rebase all the previous  commit

<img width="1728" alt="Screenshot 2024-11-23 at 6 05 00 PM" src="https://github.com/user-attachments/assets/420fa8d2-044d-4f50-bcbe-195b27a880d9">

kennethjpaul commented:
Hi,
What is the status of this PR?
Thank you
